text/0000-Element-Modifier-Managers.md

```diff
@@ -0,0 +1,444 @@
+
+This RFC is in the same spirit as the [custom components RFC](https://github.com/emberjs/rfcs/blob/master/text/0213-custom-components.md).
+
+## Detailed design
+
+This RFC introduces the concept of _modifier managers_. A modifier manager is an object that is responsible for coordinating the lifecycle events that occurs when invoking, installing and updating an element modifier.
+
+### Registering modifier managers
+
+Modifier managers are registered with the `modifier-manager` type in the
+application's registry. Similar to services, modifier managers are singleton
+objects (i.e. `{ singleton: true, instantiate: true }`), meaning that Ember
+will create and maintain (at most) one instance of each unique modifier
+manager for every application instance.
+
+To register a modifier manager, an addon will put it inside its `app` tree:
+
+```js
+// ember-basic-component/app/modifier-managers/basic.js
+
+import EmberObject from '@ember/object';
+
+export default EmberObject.extend({
+  // ...
+});
+```
+
+(Typically, the convention is for addons to define classes like this in its
+`addon` tree and then re-export them from the `app` tree. For brevity, we will
+just inline them in the `app` tree directly for the examples in this RFC.)
+
+This allows the modifier manager to participate in the DI system – receiving
+injections, using services, etc. Alternatively, component managers can also
+be registered with imperative API. This could be useful for testing or opt-ing
+out of the DI system. For example:
+
+```js
+// ember-basic-modifier/app/initializers/register-basic-modifier-manager.js
+
+const MANAGER = {
+  // ...
+};
+
+export function initialize(application) {
+  // We want to use a POJO here, so we are opt-ing out of instantiation
+  application.register('modifier-manager:basic', MANAGER, { instantiate: false });
+}
+
+export default {
+  name: 'register-basic-modifier-manager',
+  initialize
+};
+```
+
+## Determining which modifier manager to use
+
+When invoking the modifier `<p {{foo baz bar=bar}} />`, Ember will first resolve the
+modifier class (`modifier:foo`, usually the `default` export from
+`app/modifiers/foo.js`). Next, it will determine the appropiate modifier
+manager to use based on the resolved modifier class.
+
+Ember will provide a new API to assign the modifier manager for a element modifier
+class:
+
+
+```js
+// my-app/app/modifier/foo.js
+
+import EmberObject from '@ember/object';
+import { setModifierManager } from '@ember/modifier';
+
+export default setModifierManager('basic', EmberObject.extend({
```

Comment:
Yes, I think your suggestion makes sense, but would you mind chiming in over in #367?

Reply:
This seems reasonable to me, if mildly annoying for authors. Given that this is a low-level API the annoyance is probably permissible.

I think the most direct way forward is to update this RFC to include both MU and non-MU examples for how to use the `setModifierManager` API. This will help focus the number of cases and solutions, which are really not that many.